core/node/rpc/sync/client/syncer_set.go

```diff
@@ -617,18 +619,35 @@ func (ss *SyncerSet) selectNodeForStream(ctx context.Context, cookie *SyncCookie
 
 // getOrCreateSyncer returns the syncer for the given node address.
 // If the syncer does not exist, it creates a new one and starts it.
+// This implementation uses per-node-address locking to avoid blocking
+// other operations while creating syncers (which can be slow due to network calls).
 func (ss *SyncerSet) getOrCreateSyncer(nodeAddress common.Address) (StreamsSyncer, error) {
-	ss.muSyncers.Lock()
-	defer ss.muSyncers.Unlock()
+	// Check if stopped before creating
+	if ss.stopped.Load() {
+		return nil, RiverError(Err_CANCELED, "Sync stopped")
+	}
+
+	syncer, ok := ss.syncers.LoadOrStore(nodeAddress, &syncerWithLock{})
+	if ok {
+		// Entry exists, wait for initialization to complete
+		syncer.Lock()
+		defer syncer.Unlock()
+
+		// Check if initialization completed successfully
+		if syncer.StreamsSyncer != nil {
+			return syncer.StreamsSyncer, nil
+		}
 
-	if syncer, found := ss.syncers[nodeAddress]; found {
-		return syncer, nil
+		// Initialization failed, return error
+		return nil, RiverError(Err_INTERNAL, "Syncer initialization failed")
 	}
 
-	var syncer StreamsSyncer
+	// Lock the given syncer until it is initialized
+	syncer.Lock()
```

Comment:
- 2 go-routines call `getOrCreateSyncer` at the same time and request a syncer that isn't initialised.
- They end up with the same syncer instance on line 630.
- Both try to grab the lock on line 646.
- One succeeds and initialises the syncer and releases the lock on leaving the function.
- The second go-routine now grabs the lock and initialises the syncer again overwriting the previous syncer.